### ISSUE TITLE
rustdoc: render problems for re-exported implementors

### ISSUE THREAD

**Issue by jsha** [rust-lang/rust#85418]
On both stable and nightly, the "Implementors" section of a trait page incorrectly renders re-exported types. For instance, on https://doc.rust-lang.org/std/marker/trait.Sync.html, there are two entries for `LinkedList`.

### std::collections::LinkedList

![image](https://user-images.githubusercontent.com/220205/118538726-6c9dd500-b703-11eb-91e4-4ec02b144830.png)

### alloc::collections::LinkedList

![image](https://user-images.githubusercontent.com/220205/118538854-8e975780-b703-11eb-984d-973c32e4fbdb.png)

These are different in the way they are displayed: for instance, the `where` clause is displayed differently even though these are the same type. Also, the `alloc` version is missing its `[src]` link.

Looking at the HTML, the `alloc` version looks like this (I included some of the HTML of the previous, correct element for comparison):

<img width=400 src="https://user-images.githubusercontent.com/220205/118539050-cc947b80-b703-11eb-9a89-c182640df322.png">

Some issues here:

 - There's no need for a table layout
 - This should have `id=impl-Sync-51`
 - There's a `<code>` tag inside another `<code>` tag, which is unnecessary.

**Comment by jyn514** [comment]
@jsha does this also happen for re-exports in the same crate? or only across crates?

**Comment by cynecx** [comment]
This may be related https://github.com/rust-lang/rust/issues/84579#issuecomment-830863573?

**Comment by jsha** [comment]
I'm not sure, sorry!

**Comment by jsha** [comment]
It turns out the implementors that are rendered differently are inserted by JavaScript here:

https://github.com/rust-lang/rust/blob/3f9646da57946ced0b494c3c57ebffdf67136fd9/src/librustdoc/html/static/main.js#L661-L725

That pulls in a list of implementors from a .js file, which is generated by the Rust code here:

https://github.com/rust-lang/rust/blob/3f9646da57946ced0b494c3c57ebffdf67136fd9/src/librustdoc/html/render/write_shared.rs#L541-L552

I haven't figured out what triggers the HTML page to load that extra JS.

**Comment by jsha** [comment]
Here's the code that adds the script tag to load the extra JS: https://github.com/rust-lang/rust/blob/3f9646da57946ced0b494c3c57ebffdf67136fd9/src/librustdoc/html/render/print_item.rs#L755-L769

**Comment by jsha** [comment]
So, a summary of the problem here:

For in-crate implementors, rustdoc renders HTML directly. For out-of-crate implementors, rustdoc renders HTML into a string in a JS file. That means we have two different code paths putting the implementors into the HTML page so they're likely to drift apart.

This also means that out-of-crate implementors can't be sorted alongside the in-crate implementors.

The ideal would be for the out-of-crate implementors to be rendered directly into the HTML (and sorted along with all the in-crate implementors). I think that's not really practical, because as I understand it, the out-of-crate implementors can be generated by separate invocations of rustdoc.

One hack to improve the incorrect sorting would be to have a separate heading for out-of-crate implementors. But that doesn't address the "likely to drift apart" problem.

**Comment by jsha** [comment]
Relatedly, there is sometimes a section for "Implementations on Foreign Types" (e.g. https://doc.rust-lang.org/nightly/std/cmp/trait.Eq.html#foreign-impls). It's not clear to me what causes an impl to wind up in "Implementations on Foreign Types" vs appended to the Implementors section by way of JS. Maybe it makes sense for the JS-provided impls to wind up under that section?

Also we should probably define "Foreign Types." AFAICT by searching Google and The Book, this is not defined in commonly used Rust materials.

**Comment by jyn514** [comment]
"foreign types" is common in the compiler, but maybe not in the documentation. It means types from another crate; I see it used a lot in explanations of the orphan rules.

Yes, I agree putting everything in "Foreign Types" instead of JS is a better approach. I don't know enough about the code to help, sorry.

**Comment by jsha** [comment]
Unfortunately, this isn't an alternative to loading the implementations from JS, it's just an informational-organization thing.

For a trait in crate `foo`, there are two types of implementations on foreign types (I figured this out since my last post):

 - implementations inside crate `foo` (usually in the same file as the trait is defined). These are currently rendered into "Implementations on Foreign Types" during the pass that renders `foo`.
  - implementations in the foreign crates themselves. These are generated into JS when those crates are rendered. Currently, main.js will insert them under the general "Implementations" heading, but they should probably go under "Implementations on Foreign Types" for consistency.

**Comment by jyn514** [comment]
I don't think those are the same case. "Implementation on Foreign Types" is an implementation by the trait author on types from other crates". "Implementations" is an implementation on a new *downstream* crate in the same crate that the type is defined. "Foreign" here is foreign from the perspective of the impl, not the trait - it matters for coherence.

**Comment by jsha** [comment]
Right, they're not the same case from the trait author's perspective (or the struct/enum author's perspective). But from the perspective of a user reading the docs for trait Foo, is it different? As a concrete example, say trait `colors::Green` has a foreign implementation on `fruits::Apple`, and also `vegetables::Celery` has a local implementation of Green. If I'm on `Green`'s page, looking for things that implement it, how does the difference between `Apple` and `Celery` matter to me?

**Comment by jsha** [comment]
That said, the problem of sorting would actually be much _easier_ to solve by further subdividing things, so we'd have:

Implementations in this Crate on Types in this Crate
Implementations in this Crate on Types in Other Crates
Implementations in Other Crates

Then each section could be sorted individually. The "Implementations in Other Crates" section would come purely from JS (/JSON) and could be sorted in JS.

**Comment by jsha** [comment]
Related: #86632
